core/src/main/java/org/geysermc/geyser/network/GeyserNetworkManager.java

```diff
@@ -0,0 +1,221 @@
+        return this.createMessage0(channel, definition -> definition.createBuffer(data));
+    }
+
+    public <T extends MessageBuffer> Message<T> createMessage(@NonNull NetworkChannel channel, @NonNull T buffer) {
+        return this.createMessage0(channel, def -> buffer);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends MessageBuffer> Message<T> createMessage0(@NonNull NetworkChannel channel, @NonNull Function<MessageDefinition<T>, T> creator) {
+        MessageDefinition<T> definition = (MessageDefinition<T>) this.definitions.get(channel);
+        if (definition == null) {
+            throw new IllegalArgumentException("No message definition registered for channel: " + channel);
+        }
+
+        T buffer = creator.apply(definition);
+        Message<T> message = definition.createMessage(buffer);
+        if (message instanceof BedrockPacketMessage packetMessage) {
+            packetMessage.postProcess(this.session, (ByteBufMessageBuffer) buffer);
+        }
+        return message;
+    }
+
+    public PacketChannel getPacketChannel(int packetId) {
+        return this.packetChannels.get(packetId);
+    }
+
+    @SuppressWarnings("unchecked")
+    public boolean handlePacket(BedrockPacket packet, MessageDirection direction) {
+        if (this.packetChannels.isEmpty()) {
+            return true; // Avoid processing anything if we have nothing to handle
+        }
+
+        BedrockCodec codec = this.session.getUpstream().getSession().getCodec();
+        BedrockPacketDefinition<BedrockPacket> definition = codec.getPacketDefinition((Class<BedrockPacket>) packet.getClass());
+        PacketChannel channel = this.getPacketChannel(definition.getId());
+        if (channel == null) {
+            return true;
+        }
+
+        Message<?> message;
+        if (channel.packetType().isInstance(packet)) {
+            message = new BedrockPacketMessage(packet);
+        } else {
+            ByteBuf buffer = Unpooled.buffer();
+            definition.getSerializer().serialize(buffer, this.session.getUpstream().getCodecHelper(), packet);
+            message = this.createMessage(channel, new ByteBufMessageBuffer(ByteBufCodec.INSTANCE_LE, buffer));
+        }
+
+        ServerReceiveNetworkMessageEvent event = new ServerReceiveNetworkMessageEvent(this.session, channel, message, direction);
+        this.session.getGeyser().eventBus().fire(event);
+
+        // If the event is canceled, we do not want to process the packet further
+        return !event.isCancelled();
+    }
+
+    private <T extends MessageBuffer> void registerMessage(@NonNull NetworkChannel channel, @NonNull MessageDefinition<T> codec) {
+        if (this.definitions.containsKey(channel)) {
+            throw new IllegalArgumentException("Channel is already registered: " + channel);
+        }
+
+        this.definitions.put(channel, codec);
+        if (channel.isPacket()) {
```

Comment:
could also be solved using an instanceof check - is there a reason for explicitly forcing declaring whether a channel is a packet channel opposed to instanceof checking?

Reply:
Small optimization - checking a boolean is cheaper than an instanceof, especially with hot code where I could see people calling this method.